File: android/sdui-core/src/main/java/com/nba/sdui/core/renderer/LayoutTokenResolver.kt
package com.nba.sdui.core.renderer

import android.content.res.Configuration
import android.util.Log
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.ProvidableCompositionLocal
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.nba.sdui.core.tokens.FormFactorMatrix
import com.nba.sdui.core.tokens.LayoutTokenRegistry
import com.nba.sdui.core.tokens.ShadowSpec
import com.nba.sdui.core.models.generated.AspectRatioEnum
import com.nba.sdui.core.models.generated.AspectRatioUnion
import com.nba.sdui.core.models.generated.LayoutScalar
import com.nba.sdui.core.models.generated.Shadow
import com.nba.sdui.core.models.generated.ShadowOrToken
import com.nba.sdui.core.models.generated.ShadowType
import com.nba.sdui.core.request.RequestEnvelopeBuilder

/**
 * Resolves [LayoutScalar] values and token references against the codegen-baked
 * [LayoutTokenRegistry]. Registry data is bundled at build time and never fetched at runtime.
 * Form-factor changes are observed through Compose and trigger recomposition-driven invalidation.
 *
 * Rules of the road:
 *   - `LayoutScalar.IntegerValue` is returned as-is (raw dp/px on the wire).
 *   - `LayoutScalar.StringValue` must start with the `token:` prefix and resolve through
 *     spacing/radius token maps selected by form factor.
 *   - Strings without the `token:` prefix log a debug diagnostic and
 *     resolve to 0 (lenient resolver — strict decode happened upstream).
 *   - Unknown token names log `token_resolver_missing` and return 0.
 */
object LayoutTokenResolver {

    private const val TAG = "LayoutTokenResolver"
    private const val TOKEN_PREFIX = "token:"

    enum class FormFactor {
        PHONE,
        TABLET,
        TV
    }

    data class TypographySpec(
        val familyRef: String,
        val weight: Int,
        val textCase: String,
        val lineHeight: Double,
        val size: Int
    )

    val LocalFormFactor: ProvidableCompositionLocal<FormFactor> = staticCompositionLocalOf { FormFactor.PHONE }

    @Composable
    fun FormFactorProvider(content: @Composable () -> Unit) {
        val configuration = LocalConfiguration.current
        val formFactor = currentFormFactor(configuration)
        CompositionLocalProvider(LocalFormFactor provides formFactor, content = content)
    }

    /** Resolve a [LayoutScalar] to a Compose [Dp]. `null` → 0.dp. */
    fun dp(
        scalar: LayoutScalar?,
        formFactor: String = RequestEnvelopeBuilder.defaultFormFactor()
    ): Dp = intValue(scalar, toFormFactor(formFactor)).dp

    /** Resolve a [LayoutScalar] to a Compose [Dp]. `null` → 0.dp. */
    fun dp(
        scalar: LayoutScalar?,
        formFactor: FormFactor
    ): Dp = intValue(scalar, formFactor).dp

    /** Resolve a [LayoutScalar] to an integer value (dp/px logical). `null` → 0. */
    fun intValue(
        scalar: LayoutScalar?,
        formFactor: String = RequestEnvelopeBuilder.defaultFormFactor()
    ): Int = intValue(scalar, toFormFactor(formFactor))

    /** Resolve a [LayoutScalar] to an integer value (dp/px logical). `null` → 0. */
    fun intValue(
        scalar: LayoutScalar?,
        formFactor: FormFactor
    ): Int {
        return when (scalar) {
            null -> 0
            is LayoutScalar.IntegerValue -> scalar.value.toInt()
            is LayoutScalar.StringValue -> resolveTokenString(scalar.value, formFactor)
        }
    }

    /**
     * Resolve [AspectRatioUnion] (`DoubleValue` or named ratio enum) to a
     * `width / height` ratio suitable for `Modifier.aspectRatio`. `null`
     * passes through so callers can keep their "only apply when non-null"
     * gating semantics.
     */
    fun aspectRatio(union: AspectRatioUnion?): Float? {
        return when (union) {
            null -> null
            is AspectRatioUnion.DoubleValue -> union.value.toFloat()
            is AspectRatioUnion.EnumValue -> when (union.value) {
                AspectRatioEnum.The11 -> 1f
                AspectRatioEnum.The169 -> 16f / 9f
                AspectRatioEnum.The219 -> 21f / 9f
                AspectRatioEnum.The32 -> 3f / 2f
                AspectRatioEnum.The43 -> 4f / 3f
            }
        }
    }

    /**
     * Derives current form factor from platform-native signals.
     *
     * Resolution precedence:
     * 1) TV uiMode
     * 2) screen width dp (`>= 600` => TABLET, otherwise PHONE)
     * 3) Fallback PHONE
     */
    fun currentFormFactor(config: Configuration?): FormFactor {
        val modeType = config?.uiMode?.and(Configuration.UI_MODE_TYPE_MASK)
        if (modeType == Configuration.UI_MODE_TYPE_TELEVISION) {
            return FormFactor.TV
        }
        val widthDp = config?.screenWidthDp ?: 0
        return if (widthDp >= 600) FormFactor.TABLET else FormFactor.PHONE
    }

    fun typography(token: String, formFactor: FormFactor): TypographySpec? {
        val tokenName = tokenName(token) ?: return null
        val variant = LayoutTokenRegistry.typographyVariants[tokenName]
        if (variant == null) {
            Log.d(TAG, "token_resolver_missing: $token")
            return null
        }

        val category = LayoutTokenRegistry.typographyCategories[variant.categoryRef]
        if (category == null) {
            Log.d(TAG, "token_resolver_missing: $token")
            return null
        }

        val resolvedSize = when (formFactor) {
            FormFactor.PHONE -> variant.size.phone
            FormFactor.TABLET -> variant.size.tablet
            FormFactor.TV -> variant.size.tv
        }

        return TypographySpec(
            familyRef = category.familyRef,
            weight = category.weight,
            textCase = category.textCase,
            lineHeight = category.lineHeight,
            size = resolvedSize
        )
    }

    fun shadowSpec(token: String): ShadowSpec? {
        val tokenName = tokenName(token) ?: return null
        val spec = LayoutTokenRegistry.shadows[tokenName]
        if (spec == null) {
            Log.d(TAG, "token_resolver_missing: $token")
        }
        return spec
    }

    fun resolveShadowOrToken(value: ShadowOrToken?): Shadow? {
        return when (value) {
            null -> null
            is ShadowOrToken.ShadowValue -> value.value
            is ShadowOrToken.StringValue -> tokenToShadow(value.value)
        }
    }

    fun resolveShadowOrTokens(values: List<ShadowOrToken>?): List<Shadow> {
        return values?.mapNotNull { resolveShadowOrToken(it) } ?: emptyList()
    }

    fun motionDuration(token: String, formFactor: FormFactor): Int? {
        val tokenName = tokenName(token) ?: return null
        val row = LayoutTokenRegistry.motionDuration[tokenName]
        if (row == null) {
            Log.d(TAG, "token_resolver_missing: $token")
            return null
        }
        return valueFor(formFactor, row)
    }

    fun motionEasing(token: String): String? {
        val tokenName = tokenName(token) ?: return null
        val easing = LayoutTokenRegistry.motionEasing[tokenName]
        if (easing == null) {
            Log.d(TAG, "token_resolver_missing: $token")
        }
        return easing
    }

    private fun resolveTokenString(wire: String, formFactor: FormFactor): Int {
        val tokenName = tokenName(wire) ?: return 0
        val spacing = LayoutTokenRegistry.spacing[tokenName]
        if (spacing != null) {
            return valueFor(formFactor, spacing)
        }

        val radius = LayoutTokenRegistry.radius[tokenName]
        if (radius != null) {
            return valueFor(formFactor, radius)
        }

        Log.d(TAG, "token_resolver_missing: $wire")
        return 0
    }

    private fun tokenToShadow(token: String): Shadow? {
        val spec = shadowSpec(token) ?: return null
        return Shadow(
            type = if (spec.type.equals("inner", ignoreCase = true)) ShadowType.Inner else ShadowType.Drop,
            color = spec.color,
            radius = spec.radius.toDouble(),
            offsetX = spec.offsetX.toDouble(),
            offsetY = spec.offsetY.toDouble()
        )
    }

    private fun tokenName(wire: String): String? {
        if (!wire.startsWith(TOKEN_PREFIX)) {
            Log.d(TAG, "token_resolver_missing: $wire")
            return null
        }
        return wire.removePrefix(TOKEN_PREFIX)
    }

    private fun toFormFactor(formFactor: String): FormFactor {
        return when (formFactor) {
            "tablet" -> FormFactor.TABLET
            "tv" -> FormFactor.TV
            else -> FormFactor.PHONE
        }
    }

    private fun valueFor(formFactor: FormFactor, row: FormFactorMatrix<Int>): Int {
        return when (formFactor) {
            FormFactor.PHONE -> row.phone
            FormFactor.TABLET -> row.tablet
            FormFactor.TV -> row.tv
        }
    }

}

File: ios/Sources/SduiCore/Rendering/LayoutTokenResolver.swift
import CoreGraphics
import Foundation
import SwiftUI
import UIKit
import os

/// Resolves layout and design tokens using `LayoutTokenRegistry`, which parses
/// the bundled `schema/*-tokens.json` resources at startup (see AGENTS.md §3.6).
/// Token data is bundled (no runtime fetch), and form-factor updates are
/// propagated through SwiftUI reactivity (`FormFactorObserver` + `Environment`).
public enum LayoutTokenResolver {

    static let logger = Logger(subsystem: "com.nba.sdui", category: "LayoutTokenResolver")
    private static let tokenPrefix = "token:"

    static var missingTokenHook: ((String) -> Void)?

    /// Resolves a layout scalar to points for the given form factor and theme.
    static func cgFloat(
        _ scalar: LayoutScalar?,
        formFactor: String = RequestEnvelope.currentFormFactor,
        theme: String = RequestEnvelope.currentTheme
    ) -> CGFloat {
        CGFloat(intValue(scalar, formFactor: formFactor, theme: theme))
    }

    /// Preserved entry point for existing callers.
    static func resolve(
        _ scalar: LayoutScalar?,
        formFactor: FormFactor = currentFormFactor(),
        theme: String = RequestEnvelope.currentTheme
    ) -> Int {
        intValue(scalar, formFactor: formFactor.rawValue, theme: theme)
    }

    static func intValue(
        _ scalar: LayoutScalar?,
        formFactor: String = RequestEnvelope.currentFormFactor,
        theme: String = RequestEnvelope.currentTheme
    ) -> Int {
        guard let scalar else { return 0 }
        switch scalar {
        case .integer(let value):
            return value
        case .string(let wire):
            return resolveTokenString(wire, formFactor: FormFactor.fromWireValue(formFactor), theme: theme)
        }
    }

    static func resolveSpacing(_ token: String, formFactor: FormFactor = currentFormFactor()) -> Int? {
        guard let name = tokenName(from: token) else { return nil }
        return value(for: LayoutTokenRegistry.spacing[name], formFactor: formFactor)
    }

    static func resolveRadius(_ token: String, formFactor: FormFactor = currentFormFactor()) -> Int? {
        guard let name = tokenName(from: token) else { return nil }
        return value(for: LayoutTokenRegistry.radius[name], formFactor: formFactor)
    }

    static func typography(_ token: String, formFactor: FormFactor = currentFormFactor()) -> TypographySpec? {
        guard let name = tokenName(from: token),
              let variant = LayoutTokenRegistry.typographyVariants[name],
              let category = LayoutTokenRegistry.typographyCategories[variant.categoryRef] else {
            return nil
        }

        let size: Int
        switch formFactor {
        case .phone:
            size = variant.size.phone
        case .tablet:
            size = variant.size.tablet
        case .tv:
            size = variant.size.tv
        }

        return TypographySpec(
            familyRef: category.familyRef,
            weight: category.weight,
            textCase: category.textCase,
            lineHeight: category.lineHeight,
            size: size
        )
    }

    /// Resolve a wire-level `TextVariant` shorthand (e.g. `"score"`, `"labelSmall"`) to a
    /// `TypographySpec` via the bundled `typography-tokens.json` registry.
    ///
    /// The variant enum on the wire is a presentational shorthand for the full token name
    /// `nba.typography.<variant>`; this method bridges the shorthand into the same
    /// registry-driven path used for `token:`-prefixed presentational fields, so renderers
    /// never need a parallel hardcoded sizing table.
    static func typographyForVariant(_ variantName: String, formFactor: FormFactor = currentFormFactor()) -> TypographySpec? {
        return typography("token:nba.typography.\(variantName)", formFactor: formFactor)
    }

    static func shadowSpec(_ token: String) -> ShadowSpec? {
        guard let name = tokenName(from: token) else { return nil }
        return LayoutTokenRegistry.shadows[name]
    }

    static func resolveShadowOrToken(_ value: ShadowOrToken?) -> Shadow? {
        guard let value else { return nil }
        switch value {
        case .shadow(let shadow):
            return shadow
        case .string(let token):
            return tokenToShadow(token)
        }
    }

    static func resolveShadowOrTokens(_ values: [ShadowOrToken]?) -> [Shadow] {
        values?.compactMap { resolveShadowOrToken($0) } ?? []
    }

    static func motionDuration(_ token: String, formFactor: FormFactor = currentFormFactor()) -> Int? {
        guard let name = tokenName(from: token) else { return nil }
        return value(for: LayoutTokenRegistry.motionDuration[name], formFactor: formFactor)
    }

    static func motionEasing(_ token: String) -> String? {
        guard let name = tokenName(from: token) else { return nil }
        return LayoutTokenRegistry.motionEasing[name]
    }

    static func aspectRatio(_ union: AspectRatioUnion?) -> CGFloat? {
        guard let union else { return nil }
        switch union {
        case .double(let value):
            return CGFloat(value)
        case .enumeration(let value):
            switch value {
            case .the169: return 16.0 / 9.0
            case .the43: return 4.0 / 3.0
            case .the11: return 1.0
            case .the32: return 3.0 / 2.0
            case .the219: return 21.0 / 9.0
            }
        }
    }

    public static func currentFormFactor() -> FormFactor {
#if os(tvOS)
        return .tv
#elseif targetEnvironment(macCatalyst)
        return .tablet
#else
        let traits = activeTraitCollection()
        _ = UIDevice.current.orientation
        switch (traits.horizontalSizeClass, traits.verticalSizeClass) {
        case (.compact, .regular):
            return .phone
        case (.regular, .regular):
            return .tablet
        default:
            return .phone
        }
#endif
    }

    private static func resolveTokenString(_ wire: String, formFactor: FormFactor, theme: String) -> Int {
        _ = theme

        if let spacing = resolveSpacing(wire, formFactor: formFactor) { return spacing }
        if let radius = resolveRadius(wire, formFactor: formFactor) { return radius }

        logMissingToken(wire)
        return 0
    }

    private static func tokenToShadow(_ token: String) -> Shadow? {
        guard let spec = shadowSpec(token) else { return nil }
        return Shadow(
            color: spec.color,
            offsetX: Double(spec.offsetX),
            offsetY: Double(spec.offsetY),
            radius: Double(spec.radius),
            type: spec.type.lowercased() == "inner" ? .inner : .drop
        )
    }

    private static func tokenName(from wire: String) -> String? {
        guard wire.hasPrefix(tokenPrefix) else { return nil }
        return String(wire.dropFirst(tokenPrefix.count))
    }

    private static func value(for matrix: FormFactorMatrix<Int>?, formFactor: FormFactor) -> Int? {
        guard let matrix else { return nil }
        switch formFactor {
        case .phone: return matrix.phone
        case .tablet: return matrix.tablet
        case .tv: return matrix.tv
        }
    }

    private static func logMissingToken(_ wire: String) {
        logger.debug("token_resolver_missing: \(wire, privacy: .public)")
        missingTokenHook?(wire)
    }

    private static func activeTraitCollection() -> UITraitCollection {
        let scenes = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }
        let keyWindow = scenes
            .flatMap(\.windows)
            .first(where: \.isKeyWindow)

        return keyWindow?.traitCollection ?? UIScreen.main.traitCollection
    }
}

public enum FormFactor: String, CaseIterable {
    case phone
    case tablet
    case tv

    static func fromWireValue(_ rawValue: String) -> FormFactor {
        switch rawValue {
        case FormFactor.tablet.rawValue:
            return .tablet
        case FormFactor.tv.rawValue:
            return .tv
        default:
            return .phone
        }
    }
}

public struct TypographySpec {
    public let familyRef: String
    public let weight: Int
    public let textCase: String
    public let lineHeight: Double
    public let size: Int
}

public final class FormFactorObserver: ObservableObject {
    @Published public var formFactor: FormFactor

    private var notificationObservers: [NSObjectProtocol] = []

    public init(initialFormFactor: FormFactor = LayoutTokenResolver.currentFormFactor()) {
        formFactor = initialFormFactor
        UIDevice.current.beginGeneratingDeviceOrientationNotifications()

        let center = NotificationCenter.default
        notificationObservers.append(
            center.addObserver(
                forName: UIDevice.orientationDidChangeNotification,
                object: nil,
                queue: .main
            ) { [weak self] _ in
                self?.refresh()
            }
        )
        notificationObservers.append(
            center.addObserver(
                forName: FormFactorObserver.traitCollectionDidChangeNotification,
                object: nil,
                queue: .main
            ) { [weak self] _ in
                self?.refresh()
            }
        )
    }

    deinit {
        let center = NotificationCenter.default
        notificationObservers.forEach(center.removeObserver)
        UIDevice.current.endGeneratingDeviceOrientationNotifications()
    }

    public func refresh() {
        formFactor = LayoutTokenResolver.currentFormFactor()
    }

    public static let traitCollectionDidChangeNotification = Notification.Name("com.nba.sdui.formFactor.traitsDidChange")
}

private struct FormFactorEnvironmentKey: EnvironmentKey {
    static let defaultValue: FormFactor = LayoutTokenResolver.currentFormFactor()
}

public extension EnvironmentValues {
    var formFactor: FormFactor {
        get { self[FormFactorEnvironmentKey.self] }
        set { self[FormFactorEnvironmentKey.self] = newValue }
    }
}

private struct FormFactorTraitBridge: UIViewControllerRepresentable {
    func makeUIViewController(context: Context) -> FormFactorTraitBridgeController {
        FormFactorTraitBridgeController()
    }

    func updateUIViewController(_ uiViewController: FormFactorTraitBridgeController, context: Context) {}
}

private final class FormFactorTraitBridgeController: UIViewController {
    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        super.traitCollectionDidChange(previousTraitCollection)
        NotificationCenter.default.post(name: FormFactorObserver.traitCollectionDidChangeNotification, object: nil)
    }
}

public extension View {
    func withFormFactorEnvironment(_ observer: FormFactorObserver) -> some View {
        environmentObject(observer)
            .environment(\.formFactor, observer.formFactor)
            .background(FormFactorTraitBridge().frame(width: 0, height: 0))
    }
}
